src/hotspot/share/classfile/javaClasses.cpp

```diff
@@ -1933,43 +1949,33 @@ class GetThreadSnapshotClosure : public HandshakeClosure {
     return ret;
   }
 
+private:
   void detect_locks(javaVFrame* jvf, int depth) {
     Thread* current = Thread::current();
 
-    if (depth == 0 && !_found_first_lock) {
-      // See javaVFrame::print_lock_info_on() for some other cases:
-      // "waiting to re-lock in wait", "waiting on the Class initialization monitor".
-
+    if (depth == 0 && _blocker.is_empty()) {
       // If this is the first frame and it is java.lang.Object.wait(...)
       // then print out the receiver.
       if (jvf->method()->name() == vmSymbols::wait_name() &&
           jvf->method()->method_holder()->name() == vmSymbols::java_lang_Object()) {
         StackValueCollection* locs = jvf->locals();
         if (locs->is_empty()) {
-          _locks->push(LockInfo(depth, LockInfo::WAITING_ON, nullptr));
+          _blocker = LockInfo(LockInfo::WAITING_ON, OopHandle());
         } else {
-          int type = LockInfo::WAITING_ON;
           StackValue* sv = locs->at(0);
           if (sv->type() == T_OBJECT) {
             Handle o = locs->at(0)->get_obj();
-            if (_thread_status == JavaThreadStatus::BLOCKED_ON_MONITOR_ENTER) {
-              type = LockInfo::WAITING_TO_RELOCK;;
-            }
-            _locks->push(LockInfo(depth, type, o()));
+            _blocker = LockInfo(LockInfo::WAITING_ON, OopHandle(Universe::vm_global(), o()));
           }
         }
-        _found_first_lock = true;
       } else {
         oop park_blocker = java_lang_Thread::park_blocker(_java_thread());
         if (park_blocker != nullptr) {
-          _locks->push(LockInfo(depth, LockInfo::PARKING_TO_WAIT, park_blocker));
+          _blocker = LockInfo(LockInfo::PARKING_TO_WAIT, OopHandle(Universe::vm_global(), park_blocker));
```

Comment:
At some point then maybe this could move out of detect_locks as it's nothing to do with the javaVFrame, instead parkBlocker is just a property of a thread and PARKING_TO_WAIT can be PARK_BLOCKER.

Reply:
Added to TODO list

Reply:
Fixed in update commit